@here/harp-mapview/test/PoiRendererTest.ts

```diff
@@ -97,4 +103,107 @@ describe("PoiRenderer", function () {
+                    technique: {}
+                },
+                visible: true
+            } as TextElement;
+        };
+
+        let x = 1;
+        let y = 1;
+
+        const addRandomImageToCache = (
+            testCache: MapViewImageCache,
+            name: string,
+            load: boolean
+        ) => {
+            // Note, the images must be unique, otherwise the test will fail, because the internal
+            // caching mechanism of the ImageCache will have already loaded the image.
+            return testCache.addImage(name, `https://picsum.photos/${x++}/${y++}`, load);
+        };
+
+        it("poi is valid when in cache and loading started", async function () {
+            this.timeout(5000);
+
+            const testCache = new MapViewImageCache();
+            const testImageItem = addRandomImageToCache(testCache, testImageName, true);
+            const caches = [testCache];
+            const poiRenderer = new PoiRenderer(webGLRenderer, poiManager, caches);
+            const pointLabel = createPointLabel(testImageName);
+            const waitingForLoad = poiRenderer.prepareRender(pointLabel, mapEnv);
+            // This is false because the image is still being loaded in the background, notice the
+            // true flag passed to the call to testCache.addImage.
+            expect(waitingForLoad).false;
+
+            // Promise.resolve used to convert the type `ImageItem | Promise<ImageItem>` to
+            // `Promise<ImageItem>`.
+            await Promise.resolve(testImageItem);
+
+            const imageLoaded = poiRenderer.prepareRender(pointLabel, mapEnv);
+            expect(imageLoaded).true;
+
+            // Check that a second attempt to prepareRender succeeds.
+            const imageLoadedAgain = poiRenderer.prepareRender(pointLabel, mapEnv);
+            expect(imageLoadedAgain).true;
+        });
+
+        it("poi is valid when in cache and loading not started", async function () {
+            this.timeout(5000);
+
+            const testCache = new MapViewImageCache();
+            // This is the difference to the test above, we don't start loading, we expect the
+            // prepareRender function to do this.
+            addRandomImageToCache(testCache, testImageName, false);
+            const caches = [testCache];
+            const poiRenderer = new PoiRenderer(webGLRenderer, poiManager, caches);
+            const pointLabel = createPointLabel(testImageName);
+            const waitingForLoad = poiRenderer.prepareRender(pointLabel, mapEnv);
+            const result = await Promise.resolve(waitingForLoad);
+            expect(result).true;
+
+            // Check that a second attempt to prepareRender succeeds.
+            const imageLoadedAgain = poiRenderer.prepareRender(pointLabel, mapEnv);
+            expect(imageLoadedAgain).true;
+        });
+
+        it("poi is invalid when not in cache, adding to cache doesn't help", async function () {
```

Comment:
Ditto, this test is saying that the Venues issue is not solved (an image not found because the theme is still being loaded). Will we solve that on JSAPI side?

Reply:
There are two things, an item not being in the cache and an item in the cache not being loaded.

The first thing will cause the test to fail, the second won't, because it will be loaded.